tests/conftest.py

```diff
@@ -127,17 +168,67 @@ def cert_encrypted_private_key_file(example_cert):
 
 
 @pytest.fixture
-async def server():
+def restart_queue():
+    return asyncio.Queue()
+
+
+@pytest.fixture
+def restart_server(restart_queue, backend):
+    """Restart the running server from an async test function."""
+
+    async def asyncio_restart():
+        await restart_queue.put(None)
+        await restart_queue.get()
+
+    if isinstance(backend, AsyncioBackend):
+        return asyncio_restart
+
+    async def restart():
+        await backend.run_in_threadpool(AsyncioBackend().run, asyncio_restart)
+
+    return restart
+
+
+async def watch_restarts(server, queue):
+    # We can't let async tests run shutdown()/startup() directly because:
+    # * They may be running in a different async environment (e.g. trio).
+    # * When that is the case, that environment runs in a separate thread.
+    # * But shutdown/startup() needs to run on the loop from which the server was
+    # spawned, i.e. the one we're currently in.
+    # * Even if we launched `AsyncioBackend().run` in the threadpool, that would run
+    # on a *new event loop - in yet another thread!
+    # So we use a queue as a thread-safe communication channel with async tests.
+
+    while True:
+        try:
+            await queue.get()
```

Comment:
Note to self: this can’t work, I need to use get_nowait() instead. (But I’m wondering how the call to asyncio.wait() doesn’t block — currently this task should never terminate.)